<a href="https://colab.research.google.com/github/sw01-prog/python_coding_practice/blob/main/challenge_set_practice.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Fibonacci series

In [ ]:
n=int(input())

7


In [ ]:
def fibonacci(n):
 if n==1 or n==2:
  return 1
 else:
  return fibonacci(n-1)+fibonacci(n-2)

print(fibonacci(n))

13


Sum of even and odd numbers

In [ ]:
l1=[1,2,3,4,5,6,7,8]
even_sum =0
odd_sum = 0

def sum_even_odd(l1):
 even_sum =0
 odd_sum = 0
 for i in l1:
  if i%2==0:
   print(even_sum+=i)


  else:
   print(odd_sum+=i)


sum_even_odd(l1) # call the function to execute it

SyntaxError: invalid syntax (ipython-input-2029275473.py, line 10)

In [ ]:
l1=[1,2,3,4,5,6,7,8]
#even_sum =0
#odd_sum = 0

def sum_even_odd(l1):
 even_sum =0
 odd_sum = 0
 for i in l1:
  if i%2==0:
   even_sum+=i
  elif i%2!=0:
   odd_sum+=i

 return(even_sum,odd_sum)

even_sum, odd_sum = sum_even_odd(l1)
print(even_sum, odd_sum)


UnboundLocalError: cannot access local variable 'odd_sum' where it is not associated with a value

the error happens because of Python’s variable scope rules.

Inside your function sum_even_odd,

When you write even_sum += i or odd_sum += i,
Python assumes even_sum and odd_sum are local variables to the function.

But you never created them inside the function, so they are “unbound” → hence UnboundLocalError.

In [ ]:
l1=[1,2,3,4,5,6,7,8]
#even_sum =0
#odd_sum = 0

def sum_even_odd(l1):
 even_sum =0
 odd_sum = 0
 for i in l1:
  if i%2==0:
   even_sum+=i
  elif i%2!=0:
   odd_sum+=i

 return(even_sum,odd_sum)

even_sum, odd_sum = sum_even_odd(l1)
print(even_sum, odd_sum)

20 16


Palindrome number

In [ ]:
l1=[1,2,1]

def check_palindrome(l1):
 string_m= ''.join(l1)
 if string_m == string_m[::-1]:
  print('Palindrome')
 else:
  print('Not Palindrome')
check_palindrome(l1)

TypeError: sequence item 0: expected str instance, int found

The error happens because ''.join() only works on sequences of strings, but your list l1 contains integers.



In [ ]:
l1 = [1, 2, 1]

def check_palindrome(l1):
    # Convert all integers to strings first
    string_m = ''.join(map(str, l1))
    if string_m == string_m[::-1]:
        print('Palindrome')
    else:
        print('Not Palindrome')

check_palindrome(l1)


Palindrome


one more thing to check

In [ ]:
l1 = [1, 2, 1]

def check_palindrome(l1):
    # Convert all integers to strings first
    string_m = ''.join(map(str, l1))
    if string_m == string_m[::-1]:
        return True
    else:
        return False

print(check_palindrome(l1))

True


In [ ]:
l1 = [1, 2, 1]

def check_palindrome(l1):
    # Convert all integers to strings first
    string_m = ''.join(map(str, l1))
    if string_m == string_m[::-1]:
        return True
    else:
        return False

check_palindrome(l1)

True

string reverse

In [ ]:
n= input()
def reverse(n):
 rev= int(str(n)[::-1])  # int() removes leading zeros
 return rev
reverse(n)

4678


8764

Calculator based on conditions

In [ ]:
n= int(input())

def simple_calculator(n):

 for _ in range(n):
  op= int(input())

  if op==6:
   break

  if op>=1 and op<=5:
   a=int(input())
   b=int(input())

   if op==1:
    print(a+b)
   elif op==2:
    print(a-b)
   elif op==3:
    print(a*b)
   elif op==4:
    print(a//b)
   elif op==5:
    print(a%b)

  else:
   print('invalid')
   break

#simple_calculator(n)

3


In [ ]:

simple_calculator(n)

In [ ]:
def simple_calculator(n):
    for _ in range(n):
        op = int(input())  # Operation code

        if op == 6:
            break  # Exit immediately

        if 1 <= op <= 5:
            a = int(input())
            b = int(input())

            if op == 1:
                print(a + b)
            elif op == 2:
                print(a - b)
            elif op == 3:
                print(a * b)
            elif op == 4:
                print(a // b)  # Integer division
            elif op == 5:
                print(a % b)
        else:
            print("Invalid Operation")


# Example usage
n = int(input())  # Number of operations
simple_calculator(n)


1
2
4
4
0


In [ ]:
# Number of operations
n = int(input())

for _ in range(n):
    op = int(input())  # Operation code

    if op == 6:
        break  # Exit the program

    if op >= 1 and op <= 5:
        a = int(input())
        b = int(input())

        if op == 1:
            print(a + b)
        elif op == 2:
            print(a - b)
        elif op == 3:
            print(a * b)
        elif op == 4:
            print(a // b)  # Integer division
        elif op == 5:
            print(a % b)

    else:
        print("Invalid Operation")


4
1
2
4
6


KeyboardInterrupt: Interrupted by user

In [ ]:
n = int(input())
for _ in range(n):
    choice = int(input())
    if choice == 1:
        num1 = int(input())
        num2 = int(input())
        print(num1 + num2)
    elif choice == 2:
        num1 = int(input())
        num2 = int(input())
        print(num1 - num2)
    elif choice == 3:
        num1 = int(input())
        num2 = int(input())
        print(num1 * num2)
    elif choice == 4:
        num1 = int(input())
        num2 = int(input())
        # To handle potential division by zero, though constraints
        # might imply this won't happen.
        if num2 != 0:
            print(num1 // num2)
        else:
            print("Error: Division by zero")
    elif choice == 5:
        num1 = int(input())
        num2 = int(input())
        # To handle potential modulo by zero.
        if num2 != 0:
            print(num1 % num2)
        else:
            print("Error: Modulo by zero")
    elif choice == 6:
        break
    else:
        print("Invalid Operation")

3
1
2
4
6
2
1
3
-2
2
6
7
-1


Sol provided by manshu based on calculator based on condition

In [ ]:
def solution ():
   option = int(input())
   if option==6:
       return
   if option>=7:
       print("Invalid Operation")
       return solution()


   num1 = int(input())
   num2 = int(input())


   if option==1:
       print(num1+num2)
   elif option==2:
       print(num1-num2)
   elif option==3:
       print(num1*num2)
   elif option==4:
       print(num1//num2)
   elif option==5:
       print(num1%num2)


   return solution()
solution()


Convert fareinheit to celsius

In [ ]:
s= int(input())
e= int(input())
w= int(input())

def convrt_to_c(F):
 return((F-32)*5/9)

for F in range(s, e+1, w):
  print(convrt_to_c(F))

0
100
20
-17.77777777777778
-6.666666666666667
4.444444444444445
15.555555555555555
26.666666666666668
37.77777777777778


In [ ]:
S = int(input())
E = int(input())
W = int(input())

def convrt_to_c(F):
    return int((F - 32) * 5 / 9)

for F in range(S, E + 1, W):
    print(F, convrt_to_c(F))


0
100
20
0 -17
20 -6
40 4
60 15
80 26
100 37


lambda is one liner anonymous function

we defined a lambda function(upper) to convert a string to its upper case using upper().

In [ ]:
s1 = 'geeksforgeeks'
s2= lambda func: func.upper()
print(s2(s1))

GEEKSFORGEEKS


lambda with Condition Checking
- # Example: Check if a number is positive, negative, or zero

In [ ]:
z= lambda x: if x>0 print('positive') else print('negative')

SyntaxError: invalid syntax (ipython-input-1359977569.py, line 1)

In [ ]:
z= lambda x: 'positive' if x>0 else 'negative'

In [ ]:
z = lambda x: str(x)[::-1]

In [ ]:
print(z(234678))

876432


Lambda with List Comprehension

In [ ]:
z = [lambda arg=x: arg*10 for x in range(1, 5)]
for i in z:
  print(i())

10
20
30
40


In [ ]:
# Example: Check if a number is even or odd
check = lambda x: "Even" if x % 2 == 0 else "Odd"

print(check(4))
print(check(7))

Even
Odd


In [ ]:
# Example: Perform addition and multiplication in a single line
calc = lambda x, y: (x + y, x * y)

res = calc(3, 4)
print(res)

(7, 12)


if the strings has too much space between words or lettrs how to cut down

In [ ]:
text = "Hello     this    has     too   many spaces"
f = " ".join(text.split())
print(f)

Hello this has too many spaces


In [ ]:
text = "H e l l o,   W o r l d"
f= ' '.join(text.split())
print(f)

H e l l o, W o r l d


T check if a string is emty


In [ ]:
s= ''
if len(s)==0:
  print('empty')
else:
  print('not')

empty


Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4
4444
4444
4444
4444

Detailed explanation ( Input/output format, Notes, Images )
Constraints
0 <= N <= 50
Sample Input 1:
7
Sample Output 1:
7777777
7777777
7777777
7777777
7777777
7777777
7777777
Sample Input 1:
6
Sample Output 1:
666666
666666
666666
666666
666666
666666


In [2]:
n= int(input())
for i in range(n):
 print(str(n)*n)

4
4444
4444
4444
4444


Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4
*
**
***
****

Note : There are no spaces between the stars (*).
Detailed explanation ( Input/output format, Notes, Images )
Constraints
0 <= N <= 50
Sample Input 1:
5
Sample Output 1:
*
**
***
****
*****
Sample Input 2:
6
Sample Output 2:
*
**
***
****
*****
******


In [6]:
n= int(input())
for i in range(1, n+1):
  print('*'* i)

4
*
**
***
****


Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4
1
22
333
4444

Detailed explanation ( Input/output format, Notes, Images )
Constraints
0 <= N <= 50
Sample Input 1:
5
Sample Output 1:
1
22
333
4444
55555
Sample Input 2:
6
Sample Output 2:
1
22
333
4444
55555
666666


In [7]:
n= int(input())
for i in range(1, n+1):
  print(str(i)*i)

4
1
22
333
4444


Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4
1
11
121
1221

Detailed explanation ( Input/output format, Notes, Images )
Sample Input :
5

Sample Output :
1
11
121
1221
12221


In [6]:
n= int(input())
for i in range(1, n+1):
  if i ==1:
    print(1)
  elif i ==2:
   print(11)
  else:
    print(str('1'+'2'*(i-2)+'1'))

6
1
11
121
1221
12221
122221


Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4
1
21
321
4321

Detailed explanation ( Input/output format, Notes, Images )
Constraints
0 <= N <= 50
Sample Input 1:
5
Sample Output 1:
1
21
321
4321
54321
Sample Input 2:
6
Sample Output 2:
1
21
321
4321
54321


In [2]:
n= int(input())
for i in range(1, n+1):
 if i ==1:
  print(1)
 if i ==2:
   print(i+str(i-1))

2
1


TypeError: unsupported operand type(s) for +: 'int' and 'str'

In [5]:
n= int(input())
for i in range(1, n+1):
  for j in range(i, 0, -1):
    print(j, end="")
  print()


7
1
21
321
4321
54321
654321
7654321


Step 2: What does print(j, end="") mean?
Normally, print(j) adds a newline after each number.

But end="" tells Python not to add a newline, just print numbers side by side.

But since end="" is used, it does not move to the next line after finishing the row.

So after finishing, the cursor is still at the end of the same line.



This empty print() just moves the cursor to the next line.

Without it, all rows would get printed on the same line.



Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4
A
BC
CDE
DEFG

Detailed explanation ( Input/output format, Notes, Images )
Constraints
0 <= N <= 13

Sample Input 1:
5

Sample Output 1:
A
BC
CDE
DEFG
EFGHI

Sample Input 2:
6

Sample Output 2:
A
BC
CDE
DEFG
EFGHI
FGHIJK



In [7]:
n = int(input())
for i in range(1, n+1):
 row= ""
 start= 65+ (i-1)
 for j in range(i):
  row +=chr(start+j)
 print(row)

8
A
BC
CDE
DEFG
EFGHI
FGHIJK
GHIJKLM
HIJKLMNO


🔹 Step-by-step Explanation
1. Input
python
Copy
Edit
N = int(input())
Reads the number of rows.

Example: If N = 4, we want 4 rows.

2. Outer loop (rows)
python
Copy
Edit
for i in range(1, N+1):
Runs from 1 to N.

Controls which row we are printing.

So for N=4: i = 1, 2, 3, 4.

3. Initialize row string
python
Copy
Edit
row = ""
We’ll build each row as a string before printing.

4. Find the starting character of the row
python
Copy
Edit
start = 65 + (i - 1)
Recall: chr(65) = 'A'.

So 65 + (i-1) gives the ASCII code of the row’s first letter.

Example:

Row 1 → 65 + (1-1) = 65 → 'A'

Row 2 → 65 + (2-1) = 66 → 'B'

Row 3 → 65 + (3-1) = 67 → 'C'

Row 4 → 65 + (4-1) = 68 → 'D'

5. Inner loop (characters in a row)
python
Copy
Edit
for j in range(i):
    row += chr(start + j)
This loop adds characters to the row.

It runs exactly i times (so row i has i characters).

Each time, it takes the next alphabet using chr(start + j).

Example (Row 3, i=3):

start = 67 ('C')

Iterations:

j=0 → chr(67+0) = 'C'

j=1 → chr(67+1) = 'D'

j=2 → chr(67+2) = 'E'

Row string becomes "CDE".

6. Print the row
python
Copy
Edit
print(row)
After building one row, we print it.

Then outer loop moves to the next row.



Problem statement
Print the following pattern for the given number of rows.
Pattern for N = 5
E
DE
CDE
BCDE
ABCDE

Detailed explanation ( Input/output format, Notes, Images )
Constraints
0 <= N <= 26

Sample Input 1:
8

Sample Output 1:
H
GH
FGH
EFGH
DEFGH
CDEFGH
BCDEFGH
ABCDEFGH
Sample Input 2:
7
Sample Output 2:
G
FG
EFG
DEFG
CDEFG
BCDEFG
ABCDEFG



In [10]:
n= int(input())
for i in range(1, n+1):
 start = 65 + (n-i)
 end = 65+n
 row= "".join(chr(ch) for ch in range(start, end))
 print(row)

6
F
EF
DEF
CDEF
BCDEF
ABCDEF


🔹 Step-by-Step Breakdown
1. Input
python
Copy
Edit
N = int(input())
Reads the number of rows.

Example: If N = 5, we want 5 rows.

2. Outer loop
python
Copy
Edit
for i in range(1, N+1):
Runs from 1 to N.

Controls which row we are printing.

So if N=5, i = 1,2,3,4,5.

3. Find starting letter
python
Copy
Edit
start = 65 + (N - i)
chr(65) → 'A', chr(66) → 'B', …, chr(90) → 'Z'.

So 65 is ASCII code of 'A'.

For each row, we calculate the ASCII code of the first character.

Example when N=5:

Row 1 (i=1): 65 + (5-1) = 69 → 'E'

Row 2 (i=2): 65 + (5-2) = 68 → 'D'

Row 3 (i=3): 65 + (5-3) = 67 → 'C'

Row 4 (i=4): 65 + (5-4) = 66 → 'B'

Row 5 (i=5): 65 + (5-5) = 65 → 'A'

✅ That matches the expected starting letters.

4. Find ending letter
python
Copy
Edit
end = 65 + N
The row must always end at the N-th alphabet.

For N=5, that means end=70 ('F') → so range goes until just before 'F', i.e. up to 'E'.

5. Build the row
python
Copy
Edit
row = "".join(chr(ch) for ch in range(start, end))
range(start, end) → generates all ASCII values from start to end-1.

chr(ch) converts each ASCII number to a character.

"".join(...) concatenates them into a single string with no spaces.

Example (N=5, i=3 → Row 3):

start = 67 ('C'), end = 70 ('F')

Characters: 'C' 'D' 'E'

"".join(...) → "CDE"

6. Print the row
python
Copy
Edit
print(row)
After building one row string, print it.

Then go back to the outer loop for the next row.



Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4
1
11
111
1111

Detailed explanation ( Input/output format, Notes, Images )
Sample Input :
5

Sample Output :
1
11
111
1111
11111


In [12]:
n= int(input())
for i in range(1, n+1):
 if i ==1:
  print(1)
 elif i ==2:
  print(11)
 else:
  print('1'*i)

6
1
11
111
1111
11111
111111


# or

In [13]:
N = int(input())

for i in range(1, N+1):
    print("1" * i)


8
1
11
111
1111
11111
111111
1111111
11111111


Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4
1
11
202
3003

Detailed explanation ( Input/output format, Notes, Images )
Constraints:
1 <= n <= 50

Output format :
Pattern in N lines

Sample Input :
5

Sample Output :
1
11
202
3003
40004


In [14]:
n= int(input())
for i in range(1, n+1):
  if i ==1:
    print(1)
  if i ==2:
    print(11)
  else:
    print(str(i-1)+'0'*(i-2)+str(i-1))


5
1
00
11
202
3003
40004


Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4
1234
123
12
1

Detailed explanation ( Input/output format, Notes, Images )
Sample Input :
5

Sample Output :
12345
1234
123
12
1


In [ ]:
n= int(input())
for i in range(1, n+1):
  for j in range(i, 0, +1)
   print('i')

why my code didnt work because- for the no of rows which is denoted by i it contins 5 elements in the first row
- and j tells us what kind of elements to add

In [15]:
N = int(input())

for i in range(N, 0, -1):   # from N down to 1
    for j in range(1, i+1): # print numbers 1 to i
        print(j, end="")
    print()  # move to next line


4
1234
123
12
1


Print the following pattern for the given N number of rows.
Pattern for N = 3
A
BB
CCC

Detailed explanation ( Input/output format, Notes, Images )
Constraints
0 <= N <= 26
Sample Input 1:
7
Sample Output 1:
A
BB
CCC
DDDD
EEEEE
FFFFFF
GGGGGGG
Sample Input 2:
6
Sample Output 2:
A
BB
CCC
DDDD
EEEEE
FFFFFF



In [16]:
n= int(input())
for i in range(1, n+1):
  ch= chr(65+i-1)
  print(ch*i)


5
A
BB
CCC
DDDD
EEEEE


Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4
4444
333
22
1

Detailed explanation ( Input/output format, Notes, Images )
Constraints :
0 <= N <= 50

Sample Input 1:
5

Sample Output 1:
55555
4444
333
22
1

Sample Input 2:
6

Sample Output 2:
666666
55555
4444
333
22
1




In [17]:
n= int(input())
for i in range(n, 0, -1):
  print(str(i)*i)

4
4444
333
22
1


Problem statement
Print the following pattern for the given N number of rows.
Pattern for N = 4

The dots represent spaces.
Detailed explanation ( Input/output format, Notes, Images )
Constraints
0 <= N <= 50

Sample Input 1:
3

Sample Output 1:
     1
    12
  123

Sample Input 2:
4

Sample Output 2:
     1
    12
  123
1234



In [20]:
n= int(input())
for i in range(1, n+1):
 spaces = ""*(n-i)
 char= "".join(str(j) for j in range(1, i+1))
 print(spaces+char)

6
1
12
123
1234
12345
123456


Problem statement
Print the following pattern for the given number of rows.
Pattern for N = 4
1111
000
11
0

Input format : N (Total no. of rows)
Output format : Pattern in N lines
Sample Input :
5

Sample Output :
11111
0000
111
00
1



In [23]:
n= int(input())
for i in range(n):
 length= n-i
 if i%2==0:
  print('1'* length)
 else:
  print('0'*length)

6
111111
00000
1111
000
11
0


Problem statement
Print the following pattern for a given n.
For eg. N = 6
123456
  23456
    3456
      456
        56
          6
        56
      456
    3456
  23456
123456

Sample Input 1 :
4

Sample Output 1 :
1234
  234
    34
      4
    34
  234
1234
